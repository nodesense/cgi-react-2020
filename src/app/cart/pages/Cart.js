// pages/Cart.js

import React, {Component} from "react";
import PropTypes from "prop-types";

import CartList from "../components/CartList";
import CartSummary from "../components/CartSummary";

class Cart extends Component {
    static defaultProps = {
    }

    static propTypes = {
    }

    constructor(props) {
        super(props);

        this.state = {
            items: [ 
            			{id: 1, name: 'P1', price: 100, qty: 5}
            	   ],
            amount: 0, // sum of all items price * qty
            count: 0, // sum of all items qty
            flag: true
        }
    }
    
    addItem = () => {
        let id = Math.ceil(Math.random() * 10000);
        let item = {
            id,
            name: `Product ${id}`,
            price: Math.ceil(Math.random() * 100),
            qty: 1
        }

        this.setState( {
            items: [...this.state.items, item]
        })
    }
    
    // Child to parent: By means of Callback
    // Cart shall pass removeItem, updateItem functions as props to CartList
    // CartList shall pass removeItem, updateItem functions as props to CartItem
    // When -1/+1, X on events in CartItem, cartItem shall call removeItem/updateItem directly with arg
    removeItem = (id) => {
        console.log("removeItem ", id, "called by CartItem")
        
        // filter not needed the item
        this.setState((prevState, props) => {
            return {
                items: prevState.items.filter ( item => item.id !== id)
            }
        })
    }

    updateItem = (id, qty) => {
        console.log("updateItem", id, qty, "called by cartitem")
        
        this.setState( (prevState, props) => {
            return {
                // clone the array of items using map
                // clone the object/item of qty
                // if item no need to change, return as is
                // if change needed, clone it and update the qty
                items: prevState.items.map ( item => item.id !== id? item: {...item, qty: qty}  )
            }
        })
    }

    empty = () => {
        this.setState({
            items: []
        }) 
    }

    //dummy
    refresh = () => {
        this.setState({
            flag: true
        })
    }

    // derived data from state
    // make it friendly with setState functional pattern
    static recalculate(prevState, props) {
        let count = 0, 
            amount = 0;

        for (let item of prevState.items) {
            amount += item.price * item.qty;
            count += item.qty;
        }

        return {
            //es6
            amount, // amount: amount
            count, // count: count
        }
    }

    // called before render
    static getDerivedStateFromProps(props, state) {
        console.log("Cart getDerivedStateFromProps")
        // memoize, to avoid expensive computations again and again
        return  Cart.recalculate(state, props)
    }
    
    render() {
        console.log("Cart render")
        return (
            <div> 
            <h2>Cart</h2>

            <button onClick={this.addItem}>
                Add Item
            </button>


            <button onClick={this.empty}>
                Empty
            </button>

            <button onClick={this.refresh}>
                Refresh
            </button>
            

            <CartList  items={this.state.items}  
                       removeItem={this.removeItem}
                       updateItem={this.updateItem}
            />

            <CartSummary amount={this.state.amount}
                         count = {this.state.count}
            />

            </div>
        )
    }
} 




export default Cart;