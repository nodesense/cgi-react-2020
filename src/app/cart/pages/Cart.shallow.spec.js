import React from 'react';
import {shallow} from "enzyme";

import {MemoryRouter as Router} from 'react-router-dom'

import Cart from "./Cart";

describe("Cart shallow Tests", () => {

    it("test initial component", () => {
        let wrapper = shallow( 
                <Cart /> 
        )
    
        // it wont't render table rows 
        // expect(wrapper.find("tr").length).toBe(2)
        expect(wrapper).toMatchSnapshot();
            
        // instance is cart component instance
        wrapper.instance().empty()
        wrapper.update() // calls the render
        const instance = wrapper.instance()
        expect(instance.state.amount).toBe(0)
        expect(instance.state.count).toBe(0)
    })
   
})