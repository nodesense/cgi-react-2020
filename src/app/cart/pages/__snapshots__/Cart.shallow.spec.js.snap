// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart shallow Tests test initial component 1`] = `
<div>
  <h2>
    Cart
  </h2>
  <button
    onClick={[Function]}
  >
    Add Item
  </button>
  <button
    onClick={[Function]}
  >
    Empty
  </button>
  <button
    onClick={[Function]}
  >
    Refresh
  </button>
  <Link
    className="button"
    to="/checkout"
  >
    Checkout
  </Link>
  <CartList
    items={
      Array [
        Object {
          "id": 1,
          "name": "P1",
          "price": 100,
          "qty": 5,
        },
      ]
    }
    removeItem={[Function]}
    updateItem={[Function]}
  />
  <CartSummary
    amount={500}
    count={5}
  />
</div>
`;
